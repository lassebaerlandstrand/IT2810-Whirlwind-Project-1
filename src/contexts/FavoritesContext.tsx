import React, { createContext, ReactNode, useContext, useEffect, useState } from 'react';
import { Location } from '../types/api-types';

interface FavoritesContextProps {
  favorites: Location[];
  toggleFavorite: (location: Location) => void;
}

const FavoritesContext = createContext<FavoritesContextProps | undefined>(undefined);

export const useFavorites = () => {
  const context = useContext(FavoritesContext);
  if (!context) {
    throw new Error('useFavorites must be used within a FavoritesProvider');
  }
  return context;
};

export const FavoritesProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [favorites, setFavorites] = useState<Location[]>(() => {
    const storedFavorites = localStorage.getItem('favorites');
    return storedFavorites ? JSON.parse(storedFavorites) : [];
  });

  useEffect(() => {
    localStorage.setItem('favorites', JSON.stringify(favorites));
  }, [favorites]);

  const toggleFavorite = (location: Location) => {
    setFavorites((prevFavorites) => {
      const isFavorite = prevFavorites.some(
        (fav) => fav.city_name === location.city_name && fav.country_name === location.country_name,
      );
      if (isFavorite) {
        return prevFavorites.filter(
          (fav) => fav.city_name !== location.city_name || fav.country_name !== location.country_name,
        );
      } else {
        return [...prevFavorites, location];
      }
    });
  };

  return <FavoritesContext.Provider value={{ favorites, toggleFavorite }}>{children}</FavoritesContext.Provider>;
};
