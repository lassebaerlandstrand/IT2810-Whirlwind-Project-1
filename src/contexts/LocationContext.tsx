import React, { createContext, ReactNode, useContext, useState } from 'react';
import { Location } from '../types/api-types';
import LOCATIONS from '../utils/locations';

type OptionsType = {
  [key: string]: (a: Location, b: Location) => number;
};

export const options: OptionsType = {
  'City Name': (a: Location, b: Location) => a.city_name.localeCompare(b.city_name),
  'Country Name': (a: Location, b: Location) => a.country_name.localeCompare(b.country_name),
};

interface LocationContextProps {
  sortedLocations: Location[];
  setSortCondition: (sortCondition: () => (a: Location, b: Location) => number) => void;
}

const LocationContext = createContext<LocationContextProps | undefined>(undefined);

export const LocationProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [sortedLocations, setSortedLocations] = useState<Location[]>(LOCATIONS);
  const [sortCondition, setSortCondition1] = useState<() => (a: Location, b: Location) => number>(
    () => Object.values(options)[1],
  );

  // const sortLocations = useCallback(() => {
  //   let sortedData = [...LOCATIONS];
  //   if (sortCondition) {
  //     sortedData.sort(sortCondition());
  //   }
  //   setSortedLocations(sortedData);
  // }, [sortCondition]);

  // useEffect(() => {
  //   sortLocations();
  // }, [sortCondition, sortLocations]);

  const setSortCondition = () => (sortCondition: (a: Location, b: Location) => number) => {};

  return <LocationContext.Provider value={{ sortedLocations, setSortCondition }}>{children}</LocationContext.Provider>;
};

export const useLocations = () => {
  const context = useContext(LocationContext);
  if (!context) {
    throw new Error('useLocations must be used within a LocationProvider');
  }
  return context;
};
