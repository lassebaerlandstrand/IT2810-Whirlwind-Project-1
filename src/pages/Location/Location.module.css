.weatherDetailsContainer {
  display: flex;
  justify-content: space-around;
  max-width: 44rem;
  margin: 3rem auto;
  flex-wrap: wrap;
  background-color: #f7f6fc;
  padding: 1rem;
  gap: 1rem;
  border-top-left-radius: 1rem;
  border-top-right-radius: 1rem;
  list-style: none;

  li {
    width: auto;
  }
}
