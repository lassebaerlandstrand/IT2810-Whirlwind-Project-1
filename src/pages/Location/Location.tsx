import { IconCloudRain, IconWind } from '@tabler/icons-react';
import { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import Carousel from '../../components/Carousel/Carousel';
import Header from '../../components/Header/Header';
import HomeButton from '../../components/HomeButton/HomeButton';
import InfoDisplay from '../../components/InfoDisplay/InfoDisplay';
import { useWeather } from '../../hooks/useWeather';
import { WeatherInfo } from '../../types/api-types';
import LOCATIONS from '../../utils/locations';
import styles from './Location.module.css';

const translateDirection = (windDirection: number) => {
  return ['N', 'NE', 'E', 'SE', 'S', 'SW', 'W', 'NW'][Math.floor(((windDirection + 22.5) % 360) / 45)];
};

const updateInfos = (data: WeatherInfo | null) => {
  return [
    {
      icon: <IconWind />,
      infoText: 'Wind (' + (data ? translateDirection(data.wind_from_direction) : 'NA') + ')',
      amount: data?.wind_speed,
      unit: 'm/s',
      rotation: data ? data.wind_from_direction + 90 : 0,
    },

    {
      icon: <IconCloudRain />,
      infoText: 'Precipitation',
      amount: data?.precipitation_amount,
      unit: 'mm/h',
      rotation: 0,
    },
  ];
};

const Location = () => {
  const { locationName } = useParams();
  const currentIndex = LOCATIONS.findIndex(
    (location) => location.city_name?.toLowerCase() === locationName?.toLowerCase(),
  );

  const { data, isLoading, error } = useWeather(LOCATIONS[currentIndex]);

  const [infos, setInfos] = useState(updateInfos(data));

  useEffect(() => {
    setInfos(updateInfos(data));
  }, [data]);

  return (
    <main className={styles.locationContainer}>
      <Header />
      <HomeButton />
      {!isLoading && error ? (
        <>
          <p className={styles.error}>{error.name}: Something went wrong! Try reloading the page.</p>
          {console.error(error)}
        </>
      ) : (
        <>
          <Carousel currentIndex={currentIndex} data={data} />
          <InfoDisplay infos={infos} />
        </>
      )}
    </main>
  );
};

export default Location;
