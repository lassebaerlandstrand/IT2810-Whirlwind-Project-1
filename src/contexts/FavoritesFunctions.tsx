import { useContext } from 'react';
import { FavoritesContext } from './FavoritesContext';

// Had to extract to a serparate file to fix eslint warning
export const useFavorites = () => {
  const context = useContext(FavoritesContext);
  if (!context) {
    throw new Error('useFavorites must be used within a FavoritesProvider');
  }
  return context;
};
