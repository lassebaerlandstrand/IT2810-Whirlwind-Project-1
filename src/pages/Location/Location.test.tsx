import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import { MemoryRouter, Route, Routes, useParams } from 'react-router-dom';
import { vi } from 'vitest';
import { LocationProvider } from '../../contexts/LocationContext';
import LOCATIONS from '../../utils/locations';
import Location from './Location';

vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');
  return {
    ...actual,
    useParams: vi.fn(),
  };
});

const queryClient = new QueryClient();

describe('Location', () => {
  test('should match snapshot', () => {
    (useParams as jest.Mock).mockReturnValue({ locationName: LOCATIONS[0].city_name });

    const { asFragment } = render(
      <QueryClientProvider client={queryClient}>
        <MemoryRouter initialEntries={[`/location/${LOCATIONS[0].city_name}`]}>
          <LocationProvider>
            <Routes>
              <Route path="/location/:city_name" element={<Location />} />
            </Routes>
          </LocationProvider>
        </MemoryRouter>
      </QueryClientProvider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
