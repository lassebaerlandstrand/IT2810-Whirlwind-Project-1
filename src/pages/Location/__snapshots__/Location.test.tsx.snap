// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Location > should match snapshot 1`] = `
<DocumentFragment>
  <main
    class="_locationContainer_d3c209"
  >
    <header>
      <a
        class="_header_63c2ea"
        href="/"
      >
        <img
          alt="Whirlwind Logo"
          class="_img_63c2ea"
          src="/project1/src/assets/Whirlwind_Logo.png"
        />
        <div>
          <h1>
            Whirlwind
          </h1>
          <h2>
            Weather Forecast
          </h2>
        </div>
      </a>
    </header>
    <a
      class="_link_1cdd9b"
      href="/"
    >
      <svg
        class="tabler-icon tabler-icon-list _icon_1cdd9b"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9 6l11 0"
        />
        <path
          d="M9 12l11 0"
        />
        <path
          d="M9 18l11 0"
        />
        <path
          d="M5 6l0 .01"
        />
        <path
          d="M5 12l0 .01"
        />
        <path
          d="M5 18l0 .01"
        />
      </svg>
      Go back to list
    </a>
    <article
      class="_card_8ab24e _coldCard_8ab24e "
    >
      <button
        aria-label="Add New York to favorites"
        aria-pressed="false"
        class="_favoriteIcon_8ab24e"
      >
        <svg
          aria-hidden="true"
          class="tabler-icon tabler-icon-heart "
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19.5 12.572l-7.5 7.428l-7.5 -7.428a5 5 0 1 1 7.5 -6.566a5 5 0 1 1 7.5 6.572"
          />
        </svg>
      </button>
      <figure
        class="_iconContainer_8ab24e"
      >
        <svg
          aria-hidden="true"
          class="tabler-icon tabler-icon-rotate-clockwise _rotating_8ab24e _svg_8ab24e"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
          />
        </svg>
      </figure>
      <section
        class="_informationContainer_8ab24e"
      >
        <header
          class="_locationContainer_8ab24e"
        >
          <h3>
            New York
          </h3>
          <span
            aria-hidden="true"
            class="_commaSeparator_8ab24e"
          >
            , 
          </span>
          <h4>
            USA
          </h4>
        </header>
        <p
          aria-label="Temperature in New York: loading"
          class="_temperature_8ab24e _coldTemp_8ab24e"
        >
          ---Â°C
        </p>
        <p
          class="_weatherType_8ab24e"
        >
          Loading...
        </p>
      </section>
    </article>
    <div
      class="_buttonContainer_cc4f54"
    >
      <a
        aria-label="Previous city: Mawsynram"
        class="_link_cc4f54 _leftLink_cc4f54"
        href="/location/Mawsynram"
      >
        <svg
          class="tabler-icon tabler-icon-arrow-left _svg_cc4f54"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12l14 0"
          />
          <path
            d="M5 12l6 6"
          />
          <path
            d="M5 12l6 -6"
          />
        </svg>
        <h3>
          Mawsynram
        </h3>
      </a>
      <a
        aria-label="Next city: Oslo"
        class="_link_cc4f54 _rightLink_cc4f54"
        href="/location/Oslo"
      >
        <svg
          class="tabler-icon tabler-icon-arrow-right _svg_cc4f54"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12l14 0"
          />
          <path
            d="M13 18l6 -6"
          />
          <path
            d="M13 6l6 6"
          />
        </svg>
        <h3>
          Oslo
        </h3>
      </a>
    </div>
    <ul
      class="_weatherDetailsContainer_7869c2"
    >
      <li
        class="_listItemContainer_7869c2"
      >
        <article
          class="_container_7869c2"
        >
          <header
            class="_titleContainer_7869c2"
          >
            <figure
              class="_iconContainer_7869c2"
              style="transform: rotate(0deg); transition: transform 2s;"
            >
              <svg
                class="tabler-icon tabler-icon-wind "
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 8h8.5a2.5 2.5 0 1 0 -2.34 -3.24"
                />
                <path
                  d="M3 12h15.5a2.5 2.5 0 1 1 -2.34 3.24"
                />
                <path
                  d="M4 16h5.5a2.5 2.5 0 1 1 -2.34 3.24"
                />
              </svg>
            </figure>
            <h3
              class="_title_7869c2"
            >
              Wind (NA)
            </h3>
          </header>
          <section
            class="_infoContainer_7869c2"
          >
            <p
              class="_infoText_7869c2"
            >
              --.-
            </p>
            <p
              class="_infoUnit_7869c2"
            >
              m/s
            </p>
          </section>
        </article>
      </li>
      <li
        class="_listItemContainer_7869c2"
      >
        <article
          class="_container_7869c2"
        >
          <header
            class="_titleContainer_7869c2"
          >
            <figure
              class="_iconContainer_7869c2"
              style="transform: rotate(0deg); transition: transform 2s;"
            >
              <svg
                class="tabler-icon tabler-icon-cloud-rain "
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7 18a4.6 4.4 0 0 1 0 -9a5 4.5 0 0 1 11 2h1a3.5 3.5 0 0 1 0 7"
                />
                <path
                  d="M11 13v2m0 3v2m4 -5v2m0 3v2"
                />
              </svg>
            </figure>
            <h3
              class="_title_7869c2"
            >
              Precipitation
            </h3>
          </header>
          <section
            class="_infoContainer_7869c2"
          >
            <p
              class="_infoText_7869c2"
            >
              --.-
            </p>
            <p
              class="_infoUnit_7869c2"
            >
              mm/h
            </p>
          </section>
        </article>
      </li>
    </ul>
  </main>
</DocumentFragment>
`;
